File: app/src/main/java/org/blitzortung/android/data/provider/blitzortung/BlitzortungHttpDataProvider.java
package org.blitzortung.android.data.provider.blitzortung;

import android.util.Log;
import org.blitzortung.android.app.Main;
import org.blitzortung.android.data.beans.StrikeAbstract;
import org.blitzortung.android.data.beans.Station;
import org.blitzortung.android.data.beans.RasterParameters;
import org.blitzortung.android.data.provider.DataProvider;
import org.blitzortung.android.data.provider.DataProviderType;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.Authenticator;
import java.net.PasswordAuthentication;
import java.net.URL;
import java.net.URLConnection;
import java.util.*;
import java.util.zip.GZIPInputStream;

public class BlitzortungHttpDataProvider extends DataProvider {

    private UrlFormatter urlFormatter;

    private MapBuilder<StrikeAbstract> strikeMapBuilder;
    private MapBuilder<Station> stationMapBuilder;

    public enum Type {STRIKES, STATIONS}

    private class MyAuthenticator extends Authenticator {

        public PasswordAuthentication getPasswordAuthentication() {
            return new PasswordAuthentication(username, password.toCharArray());
        }
    }

    private long latestTime = 0;

    public BlitzortungHttpDataProvider() {
        this(new UrlFormatter(), new MapBuilderFactory());
    }

    public BlitzortungHttpDataProvider(UrlFormatter urlFormatter, MapBuilderFactory mapBuilderFactory) {
        this.urlFormatter = urlFormatter;
        strikeMapBuilder = mapBuilderFactory.createAbstractStrikeMapBuilder();
        stationMapBuilder = mapBuilderFactory.createStationMapBuilder();

    }

    @Override
    public List<StrikeAbstract> getStrikes(int timeInterval, int intervalOffset, int region) {

        List<StrikeAbstract> strikes = new ArrayList<StrikeAbstract>();

        TimeZone tz = TimeZone.getTimeZone("UTC");
        Calendar intervalTime = new GregorianCalendar(tz);

        if (username != null && username.length() != 0 && password != null && password.length() != 0) {

            try {
                IntervalTimer intervalTimer = new IntervalTimer(10 * 60 * 1000l);
                long startTime = System.currentTimeMillis() - timeInterval * 60 * 1000;

                intervalTimer.startInterval(Math.max(latestTime, startTime));

                while (intervalTimer.hasNext()) {
                    intervalTime.setTimeInMillis(intervalTimer.next());

                    BufferedReader reader = readFromUrl(Type.STRIKES, region, intervalTime);
                    if (reader == null) {
                        continue;
                    }

                    int size = 0;
                    String line;
                    while ((line = reader.readLine()) != null) {
                        size += line.length();

                        StrikeAbstract strike = strikeMapBuilder.buildFromLine(line);
                        long timestamp = strike.getTimestamp();

                        if (timestamp > latestTime && timestamp >= startTime) {
                            strikes.add(strike);
                        }
                    }
                    Log.v(Main.LOG_TAG,
                            String.format("BliztortungHttpDataProvider: read %d bytes (%d new strikes) from region %d", size, strikes.size(), region));

                    reader.close();
                }

                if (strikes.size() > 0) {
                    latestTime = strikes.get(strikes.size() - 1).getTimestamp();
                }

            } catch (Exception e) {
                throw new RuntimeException(e);
            }

        } else {
            throw new RuntimeException("no credentials provided");
        }

        return strikes;
    }

    public boolean returnsIncrementalData() {
        return latestTime != 0;
    }

    private BufferedReader readFromUrl(Type type, int region) {
        return readFromUrl(type, region, null);
    }

    private BufferedReader readFromUrl(Type type, int region, Calendar intervalTime) {

        boolean useGzipCompression = false;

        if (type == Type.STATIONS) {
            useGzipCompression = true;
        }

        Authenticator.setDefault(new MyAuthenticator());

        BufferedReader reader;

        String urlString = urlFormatter.getUrlFor(type, region, intervalTime, useGzipCompression);
        try {
            URL url;
            url = new URL(urlString);
            URLConnection connection = url.openConnection();
            connection.setConnectTimeout(60000);
            connection.setReadTimeout(60000);
            connection.setAllowUserInteraction(false);
            InputStream ins = connection.getInputStream();
            if (useGzipCompression) {
                ins = new GZIPInputStream(ins);
            }
            reader = new BufferedReader(new InputStreamReader(ins));
        } catch (FileNotFoundException e) {
            Log.w(Main.LOG_TAG, String.format("URL '%s' not found", urlString));
            return null;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        return reader;
    }

    @Override
    public List<Station> getStations(int region) {
        List<Station> stations = new ArrayList<Station>();

        if (username != null && username.length() != 0 && password != null && password.length() != 0) {

            try {
                BufferedReader reader = readFromUrl(Type.STATIONS, region);

                int size = 0;
                String line;
                while ((line = reader.readLine()) != null) {
                    size += line.length();
                    try {
                        Station station = stationMapBuilder.buildFromLine(line);
                        stations.add(station);
                    } catch (NumberFormatException e) {
                        Log.w(Main.LOG_TAG, String.format("BlitzortungHttpProvider: error parsing '%s'", line));
                    }
                }
                Log.v(Main.LOG_TAG,
                        String.format("BlitzortungHttpProvider: read %d bytes (%d stations) from region %d", size, stations.size(), region));

                reader.close();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }

        } else {
            throw new RuntimeException("no credentials provided");
        }

        return stations;
    }

    @Override
    public DataProviderType getType() {
        return DataProviderType.HTTP;
    }

    @Override
    public void setUp() {
    }

    @Override
    public void shutDown() {
    }

    @Override
    public int[] getHistogram() {
        return null;
    }

    @Override
    public RasterParameters getRasterParameters() {
        return null;
    }

    @Override
    public List<StrikeAbstract> getStrikesGrid(int intervalDuration, int intervalOffset, int rasterSize, int countThreshold, int region) {
        return null;
    }

    @Override
    public void reset() {
        latestTime = 0;
    }

    @Override
    public boolean isCapableOfHistoricalData() {
        return false;
    }

}

File: app/src/main/java/org/blitzortung/android/data/provider/blitzortung/BlitzortungHttpDataProvider.kt
/*

   Copyright 2015-2016 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.data.provider.blitzortung

import android.content.SharedPreferences
import android.util.Log
import org.blitzortung.android.app.Main
import org.blitzortung.android.app.view.OnSharedPreferenceChangeListener
import org.blitzortung.android.app.view.PreferenceKey
import org.blitzortung.android.app.view.get
import org.blitzortung.android.data.Flags
import org.blitzortung.android.data.History
import org.blitzortung.android.data.Parameters
import org.blitzortung.android.data.beans.Station
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.data.provider.DataProviderType
import org.blitzortung.android.data.provider.data.DataProvider
import org.blitzortung.android.data.provider.data.DataProvider.DataRetriever
import org.blitzortung.android.data.provider.data.initializeResult
import org.blitzortung.android.data.provider.result.ResultEvent
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.FileNotFoundException
import java.io.InputStreamReader
import java.net.Authenticator
import java.net.PasswordAuthentication
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.URL
import java.util.Calendar
import java.util.GregorianCalendar
import java.util.TimeZone
import java.util.zip.GZIPInputStream
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max

@Singleton
class BlitzortungHttpDataProvider @Inject constructor(
    preferences: SharedPreferences,
    private val urlFormatter: UrlFormatter,
    mapBuilderFactory: MapBuilderFactory
) : OnSharedPreferenceChangeListener, DataProvider {

    private val strikeMapBuilder: MapBuilder<Strike> = mapBuilderFactory.createStrikeMapBuilder()
    private val stationMapBuilder: MapBuilder<Station> = mapBuilderFactory.createStationMapBuilder()
    private var latestTime: Long = 0
    private var strikes: List<Strike> = emptyList()
    private var parameters: Parameters? = null

    private lateinit var username: String
    private lateinit var password: String

    init {
        preferences.registerOnSharedPreferenceChangeListener(this)
        onSharedPreferenceChanged(preferences, PreferenceKey.USERNAME, PreferenceKey.PASSWORD)
    }

    private fun readFromUrl(type: Type, region: Int, intervalTime: Calendar? = null): BufferedReader {

        val useGzipCompression = type == Type.STATIONS

        val reader: BufferedReader

        val urlString = urlFormatter.getUrlFor(type, region, intervalTime, useGzipCompression)

        try {
            val url = URL(urlString)
            val connection = url.openConnection()
            connection.connectTimeout = 40000
            connection.readTimeout = 40000
            connection.allowUserInteraction = false
            var inputStream = connection.inputStream
            if (useGzipCompression) {
                inputStream = GZIPInputStream(inputStream)
            }

            reader = inputStream.bufferedReader()
        } catch (e: Exception) {
            when (e) {
                is FileNotFoundException -> {
                    Log.w(Main.LOG_TAG, "BlitzortungHttpDataProvider.readFromUrl() $urlString not found")
                    return BufferedReader(InputStreamReader(ByteArrayInputStream("".toByteArray())))
                }

                else -> {
                    Log.w(Main.LOG_TAG, "BlitzortungHttpDataProvider.readFromUrl() $urlString failed")
                    throw RuntimeException(e)
                }
            }
        }

        Log.v(Main.LOG_TAG, "BlitzortungHttpDataProvider.readFromUrl() $urlString")

        return reader
    }

    /**
     * Used to retrieve Data from Blitzortung
     * @param readerSeq A sequence of nullable BufferedReader, which the data is read from
     * @param parse A Lambda which receives a sequence of lines from a buffered reader and transforms them into a sequence of T
     * @return Returns a list of parsed T's
     */
    private fun <T : Any> retrieveData(
        logMessage: String,
        readerSeq: Sequence<BufferedReader?>,
        parse: (String) -> T?
    ): List<T> {
        var size = 0

        val strikeSequence: Sequence<T> = readerSeq.filterNotNull().flatMap { reader ->
            reader.lineSequence().mapNotNull { line ->
                size += line.length

                parse(line)
            }
        }

        val strikeList = try {
            strikeSequence.toList()
        } catch (e: SocketException) {
            Log.w(Main.LOG_TAG, e)
            emptyList()
        } catch (e: SocketTimeoutException) {
            Log.w(Main.LOG_TAG, e)
            emptyList()
        }

        Log.v(Main.LOG_TAG, logMessage.format(size, strikeList.count()))

        return strikeList
    }

    override val type: DataProviderType = DataProviderType.HTTP

    override fun reset() {
        latestTime = 0L
        strikes = emptyList()
    }

    enum class Type {
        STRIKES, STATIONS
    }

    override fun <T> retrieveData(retrieve: DataRetriever.() -> T): T {
        return Retriever().retrieve()
    }

    private inner class Retriever : DataRetriever {
        override fun getStrikes(parameters: Parameters, history: History?, flags: Flags): ResultEvent {
            var result = initializeResult(parameters, history, flags)

            if (parameters != this@BlitzortungHttpDataProvider.parameters) {
                this@BlitzortungHttpDataProvider.parameters = parameters
                reset()
            }

            val intervalDuration = parameters.intervalDuration
            val intervalOffset = parameters.intervalOffset
            val region = parameters.region

            val tz = TimeZone.getTimeZone("UTC")
            val intervalTime = GregorianCalendar(tz)

            val millisecondsPerMinute = 60 * 1000L
            val endTime = System.currentTimeMillis() + intervalOffset * millisecondsPerMinute
            val startTime = endTime - intervalDuration * millisecondsPerMinute
            val intervalSequence =
                createTimestampSequence(10 * millisecondsPerMinute, max(startTime, latestTime), endTime)

            Authenticator.setDefault(MyAuthenticator())

            val strikes =
                retrieveData(
                    "BlitzortungHttpDataProvider.getStrikes() read %d bytes (%d new strikes) from region $region",
                    intervalSequence.map {
                        intervalTime.timeInMillis = it

                        return@map readFromUrl(Type.STRIKES, region, intervalTime)
                    }) { strikeMapBuilder.buildFromLine(it) }

            if (latestTime > 0L) {
                result = result.copy(updated = strikes.size)
                val expireTime = result.referenceTime - (intervalDuration - intervalOffset) * millisecondsPerMinute
                this@BlitzortungHttpDataProvider.strikes =
                    this@BlitzortungHttpDataProvider.strikes.filter { it.timestamp > expireTime }
                this@BlitzortungHttpDataProvider.strikes += strikes
            } else {
                this@BlitzortungHttpDataProvider.strikes = strikes
            }

            if (strikes.isNotEmpty()) {
                latestTime = endTime - millisecondsPerMinute
                Log.v(Main.LOG_TAG, "BlitzortungHttpDataProvider.getStrikes() set latest time to $latestTime")
            }

            result = result.copy(strikes = this@BlitzortungHttpDataProvider.strikes, referenceTime = endTime)

            return result
        }

        override fun getStrikesGrid(parameters: Parameters, history: History?, flags: Flags): ResultEvent {
            return initializeResult(parameters, history, flags)
        }

        override fun getStations(region: Int): List<Station> {

            Authenticator.setDefault(MyAuthenticator())

            return retrieveData(
                "BlitzortungHttpDataProvider.getStations() read %d bytes (%d stations) from region $region",
                sequenceOf(readFromUrl(Type.STATIONS, region))
            ) {
                stationMapBuilder.buildFromLine(it)
            }
        }
    }

    private inner class MyAuthenticator : Authenticator() {
        override fun getPasswordAuthentication(): PasswordAuthentication {
            return PasswordAuthentication(username, password.toCharArray())
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: PreferenceKey) {
        when (key) {
            PreferenceKey.USERNAME -> username = sharedPreferences.get(key, "")

            PreferenceKey.PASSWORD -> password = sharedPreferences.get(key, "")

            else -> {
            }
        }
    }
}
